sw/device/tests/dif/dif_keymgr_unittest.cc

```diff
@@ -0,0 +1,1114 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_keymgr.h"
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/testing/mock_mmio.h"
+
+#include "keymgr_regs.h"  // Generated
+
+namespace dif_keymgr_unittest {
+namespace {
+
+/**
+ * Returns a `uint32_t` with a single zero bit.
+ */
+uint32_t AllOnesExcept(uint32_t index) { return ~(1u << index); }
+
+/**
+ * Returns a vector of values for a given enum type.
+ *
+ * Assumes that the values are sequential and the first value is 0. `last` must
+ * be the last valid value for the given enum and is included in the returned
+ * vector.
+ */
+template <typename T>
+std::vector<T> CreateEnumVector(T last) {
+  using TT = typename std::underlying_type<T>::type;
+  std::vector<T> res;
+  for (TT i = 0; i <= static_cast<TT>(last); ++i) {
+    res.push_back(static_cast<T>(i));
+  }
+  return res;
+}
+
+/**
+ * Returns a seemingly valid, i.e. nonzero, pointer for the given type or a
+ * `nullptr`.
+ */
+template <typename T>
+T *GetGoodBadPtrArg(bool is_good) {
+  if (is_good) {
+    return reinterpret_cast<T *>(0x1);
+  } else {
+    return nullptr;
+  }
+}
+
+/**
+ * Returns a valid or invalid value for the given enum type.
+ *
+ * `last` must be the last valid value of the given enum type and `last+1` must
+ * be an invalid value.
+ */
+template <typename T>
+T GetGoodBadEnumArg(bool is_good, T last) {
+  if (is_good) {
+    return last;
+  } else {
+    return static_cast<T>(last + 1);
+  }
+}
+
+/**
+ * Constants used in tests.
+ */
+static constexpr std::array<uint32_t, 3> kStatesWithOperationalNextStates{
+    KEYMGR_WORKING_STATE_STATE_VALUE_INIT,
+    KEYMGR_WORKING_STATE_STATE_VALUE_CREATOR_ROOT_KEY,
+    KEYMGR_WORKING_STATE_STATE_VALUE_OWNER_INTERMEDIATE_KEY,
+};
+static constexpr std::array<uint32_t, 4> kStatesWithNonOperationalNextStates{
+    KEYMGR_WORKING_STATE_STATE_VALUE_RESET,
+    KEYMGR_WORKING_STATE_STATE_VALUE_OWNER_KEY,
+    KEYMGR_WORKING_STATE_STATE_VALUE_DISABLED,
+    KEYMGR_WORKING_STATE_STATE_VALUE_INVALID,
+};
+
+class DifKeymgrTest : public testing::Test, public mock_mmio::MmioTest {
+ protected:
+  /**
+   * Parameters for initializing a `dif_keymgr_t`.
+   */
+  const dif_keymgr_params_t params_ = {.base_addr = dev().region()};
+};
+
+/**
+ * Class for parameterizing bad argument tests for functions with two arguments.
+ */
+class BadArgsTwo : public DifKeymgrTest,
+                   public testing::WithParamInterface<std::tuple<bool, bool>> {
+ protected:
+  bool AllParamsGood() {
+    return std::get<0>(GetParam()) && std::get<1>(GetParam());
+  }
+
+  void SetUp() override {
+    if (AllParamsGood()) {
+      // Only test negative cases.
+      GTEST_SKIP();
+    }
+  }
+};
+
+INSTANTIATE_TEST_SUITE_P(
```

Comment:
Honestly, this is super clever and I like it, but the tuples... aren't awesome. But! I had a crazy idea for making it a tiny bit less awful and more type safe! Consider the following code:

```cc
template <typename First, First(first_fn)(bool), typename Second, Second(second_fn)(bool)>
class BadArgs2 : public ..., WithParamInterface<std::tuple<bool, bool>> {
  public:
    template <size_t N>
    auto MaybeBad() {
      auto fn = std::get<N>(std::make_tuple(first_fn, second_fn));
      auto flag = std::get<N>(GetParam());
      return fn(flag);
    }
  protected:
    void SetUp() { ... }
}
```

Then you'd use this as `BadArgs<T*, GetGoodBadPtrArg<T>, ...>` and call `MaybeBad<0>()` in the function body. Given how frequently we deal with this kind of thing I almost wonder if it should be a library type...

While I would prefer to have written this using `template <auto first, auto second>`, that's not available until C++17 (I continue to regret saying C++14 was ok, lol).

Reply:
I agree, it's a good idea to put this in a library, created #5804.